### PR TITLE
Feature/exception handler refactor

### PR DESCRIPTION
**Pull request by faizalrub-datastax**
**What this PR does**: Validation - Log partition range level exceptions. Exceptions file can be then used as input for validation job rerun.

**Which issue(s) this PR fixes**:
Fixes #<CDM-25>

**Checklist:**
- [ ] Automated Tests added/updated
- [x] Documentation added/updated
- [ ] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/java/datastax/cdm/job/DiffJobSession.java [hunk 1]
```diff
@@ -145,10 +145,21 @@ public void getDataAndDiff(BigInteger min, BigInteger max) {
                 logger.error("Error occurred during Attempt#: {}", attempts, e);
                 logger.error("Error with PartitionRange -- ThreadID: {} Processing min: {} max: {} -- Attempt# {}",
                         Thread.currentThread().getId(), min, max, attempts);
+                if (null != tokenRangeExceptionDir && tokenRangeExceptionDir.trim().length() > 0 && attempts == maxAttempts) {
```
Comment (msmygit) [inline src/main/java/datastax/cdm/job/DiffJobSession.java:148]:
```suggestion
                if (StringUtils.isNotBlank(tokenRangeExceptionDir) && attempts == maxAttempts) {
```

File: src/main/java/datastax/cdm/job/ExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package datastax.cdm.job;
```
Comment (msmygit) [inline src/main/java/datastax/cdm/job/ExceptionHandler.java:1]: #nitpick not related to this PR, but we should standardize on the package naming convention. Maybe `com.datastax.cdm.job`? 🤔

Comment (faizalrub-datastax) [reply src/main/java/datastax/cdm/job/ExceptionHandler.java:1]: Sounds good. Let us confirm in our Thu call.

File: src/main/java/datastax/cdm/job/DiffJobSession.java [hunk 3]
```diff
@@ -14,6 +14,7 @@
 import datastax.cdm.cql.statement.TargetSelectByPKStatement;
 import datastax.cdm.properties.ColumnsKeysTypes;
 import datastax.cdm.properties.KnownProperties;
+import org.apache.commons.lang.StringUtils;
```
Comment (msmygit) [inline src/main/java/datastax/cdm/job/DiffJobSession.java:17]:
```suggestion
import org.apache.commons.lang3.StringUtils;
```

File: src/resources/sparkConf.properties [hunk 4]
```diff
@@ -164,12 +164,15 @@ spark.cdm.schema.target.column.id.names           partition_col1,partition_col2,
...
+#                           counter will become 5323+5323 = 10646.
+# spark.tokenRange
+#   .exceptionDir         : When set partition ranges corresponding to the exceptions will be logged in
+#                           file with the name <keyspace>.<table_name> in this directory.
 #-----------------------------------------------------------------------------------------------------------
 spark.cdm.autocorrect.missing                     false
 spark.cdm.autocorrect.mismatch                    false
 #spark.cdm.autocorrect.missing.counter             false
-
+#spark.tokenRange.exceptionDir                      aaa/bbb
```
Comment (msmygit) [inline src/resources/sparkConf.properties:175]:
```suggestion
#spark.tokenRange.exceptionDir                      /path/to/exceptions_directory
```

**Review by mfmaher2 — APPROVED** [review]

File: src/main/java/com/datastax/cdm/job/DiffJobSession.java [hunk 5]
```diff
@@ -12,6 +12,7 @@
 import com.datastax.cdm.cql.statement.OriginSelectByPartitionRangeStatement;
 import com.datastax.cdm.cql.statement.TargetSelectByPKStatement;
 import com.datastax.cdm.properties.KnownProperties;
+import org.apache.commons.lang.StringUtils;
```
Comment (msmygit) [inline src/main/java/com/datastax/cdm/job/DiffJobSession.java:15]:
```suggestion
import org.apache.commons.lang3.StringUtils;
```

File: src/main/java/com/datastax/cdm/properties/KnownProperties.java [hunk 6]
```diff
@@ -132,6 +132,15 @@ public enum PropertyType {
         defaults.put(MAX_RETRIES, "0");
     }
 
+    //==========================================================================
+    // Error handling
+    //==========================================================================
+    public static final String TOKEN_RANGE_EXCEPTION_DIR = "spark.tokenRange.exceptionDir"; // aaa/bbb
+    public static final String TOKEN_RANGE_EXCEPTION_FILE = "spark.input.partitionFile"; // aaa/bbb/filename
+    static {
+            types.put(TOKEN_RANGE_EXCEPTION_DIR, PropertyType.STRING);
```
Comment (pravinbhat) [inline src/main/java/com/datastax/cdm/properties/KnownProperties.java:141]: Can we combine these two properties into one to reduce properties? The code should be able to easily figure the `dir` part from the full file-path.

**Comment by msmygit** [comment]
@faizalrub-datastax do we need these changes? If so, could you please get the merge conflicts fixed so we could get these merged into the main branch? TY!

File: src/main/java/com/datastax/cdm/properties/KnownProperties.java (hunk 6 above)
Comment (faizalrub-datastax) [reply src/main/java/com/datastax/cdm/properties/KnownProperties.java:146]: TOKEN_RANGE_EXCEPTION_FILE is removed now. Refactored to use the exiting class DiffPartitionsFromFile.scala for reading partition ranges for diff exceptions as well.

File: SIT/smoke/01_basic_kvp/cdm.sh [hunk 7]
```diff
@@ -21,4 +21,4 @@ EOF
 spark-submit \
   --properties-file /smoke/01_basic_kvp/migrate.properties \
   --master "local[*]" \
-  --class datastax.astra.migrate.DiffData /local/cassandra-data-migrator.jar
+  --class datastax.cdm.job.DiffData /local/cassandra-data-migrator.jar
```
Comment (msmygit) [inline SIT/smoke/01_basic_kvp/cdm.sh:24]: shouldn't this be `com.datastax.cdm.job.DiffData` as the class is [here](https://github.com/datastax/cassandra-data-migrator/blob/main/src/main/scala/com/datastax/cdm/job/DiffData.scala)?

File: SIT/smoke/01_basic_kvp/cdm.sh [hunk 8]
```diff
@@ -10,7 +10,7 @@ EOF
 spark-submit \
   --properties-file /smoke/01_basic_kvp/migrate.properties \
   --master "local[*]" \
-  --class datastax.astra.migrate.Migrate /local/cassandra-data-migrator.jar 
+  --class datastax.cdm.job.Migrate /local/cassandra-data-migrator.jar
```
Comment (msmygit) [inline SIT/smoke/01_basic_kvp/cdm.sh:13]: shouldn't this be `com.datastax.cdm.job.Migrate` as the class is [here](https://github.com/datastax/cassandra-data-migrator/blob/main/src/main/scala/com/datastax/cdm/job/Migrate.scala)?

Comment (faizalrub-datastax) [reply SIT/smoke/01_basic_kvp/cdm.sh:13]: @msmygit  It should be --class com.datastax.cdm.job cc: @mieslep , will raise a separate JIRA ticket to resolve the refactoring issues  synk security check.

File: SIT/smoke/01_basic_kvp/cdm.sh (hunk 7 above)
Comment (faizalrub-datastax) [reply SIT/smoke/01_basic_kvp/cdm.sh:24]: It should be --class com.datastax.cdm.job cc: @mieslep , will raise a separate JIRA ticket to resolve the refactoring issues  synk security check.

**Pull request merged**